052-k8s-server-side-apply.md

```diff
@@ -0,0 +1,77 @@
+
+It appears as simple as modifying [this line](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNamespacedResourceOperator.java#LL263C88-L263C102).
+From:
+
+```
+....patch(PatchContext.of(PatchType.JSON), desired);
+```
+
+To:
+
+```
+....patch(new PatchContext.Builder().withPatchType(PatchType.SERVER_SIDE_APPLY).withForce(true).build(), desired);
+```
+
+client.services().withName("name").patch(service);
+
+And Kubernetes recommend it is used on CREATE also:
+[From](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNamespacedResourceOperator.java#L272):
+
+```
+ReconcileResult<T> result = ReconcileResult.created(operation().inNamespace(namespace).resource(desired).create());
+```
+
+To:
+
+```
+ReconcileResult<T> result = ReconcileResult.patched(operation().inNamespace(namespace).withName(name).patch(new PatchContext.Builder().withPatchType(PatchType.SERVER_SIDE_APPLY).withForce(true).build(), desired);
+```
+
+According to the [package being used](https://github.com/fabric8io/kubernetes-client/blob/v6.5.1/doc/CHEATSHEET.md#server-side-apply).
+
+NOTE: The version of fabric8io being used is [6.5.1](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/pom.xml#L106) which [contains Server Side Apply](https://github.com/fabric8io/kubernetes-client/blob/v6.5.1/doc/CHEATSHEET.md#server-side-apply).
+
+It is likely that other parts of the codebase will need modifying in terms of label/annotation discovery so that they only check labels/annotations that are owned by themselves.
+
+The [`ResourceDiff`](https://github.com/strimzi/strimzi-kafka-operator/blob/c3522cf4b17004004a676854d37ba01bb9a44800/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ResourceDiff.java#L46-L78) will need to filter only the parts of the resource that Strimzi owns as to not trigger a reconcile loop for externally owned resource paths.
+This will allow the removal of `ignorableFields` throughout the codebase as paths are being selected now, rather than filtering out fields.
```

Comment:
You say we need to filter things. What will we filter them against if we remove the `ignorableFields`?

Reply:
Currently `ResourceDiff` is looking at two things - the current state, and desired state. It's looking at all differences, and filtering out certain things explicitly that it knows other people manage (but is breaking whenever anything else is changed, even though it's not something it should care about, like additional labels etc)

Instead, it should look at the differences that are related to the desired state - instead of it being a filter of "is this in the ignored paths", it's a filter of "is this path in the desired state"

We don't need to worry about things that have gone away from the desired state, because the managed fields on the server side will handle it. There's no need to look at the managed fields on the client side here

Reply:
> even though it's not something it should care about, like additional labels

I think you are misunderstanding the design. Not every unset field means we do not care about this field. It means that at this point we are fine with its default. And for the labels and annotations, you are actually proposing here a change to the semantic which today is that for example the labels and annotations set through Strimzi should be the only ones present on the resource => i.es any other labels and annotations not defined there should be removed.

That means that it needs to be clear from the proposal how would it work for the users who are fine with the current design and how it impacts the code and the usability. Right now, I'm not really sure what the actual impact is in some of these areas.

Reply:
> I think you are misunderstanding the design. Not every unset field means we do not care about this field. It means that at this point we are fine with its default. 

Mhm. Well, I think what we're saying here is we think this design is fundamentally wrong, and it should be changed, to allow for standard Kubernetes API machinery to function as expected. Strimzi should work like other operators, which ignore changes to fields they're not explicitly setting.

> which today is that for example the labels and annotations set through Strimzi should be the only ones present on the resource => i.es any other labels and annotations not defined there should be removed.

This is not true. Annotations starting with `cattle.io/` or `field.cattle.io` are silently ignored. This is not mentioned anywhere in the Strimzi API documentation

> That means that it needs to be clear from the proposal how would it work for the users who are fine with the current design and how it impacts the code and the usability. Right now, I'm not really sure what the actual impact is in some of these areas.

Fair. Right now, anyone who's happy with the current design is clearly not suffering from a tight reconcilation loop - which means they don't have anything that is fighting with Strimzi. This means that in this new proposal, there would be no change, as there's nothing they have deployed trying to make such changes.

I think the impact on the code and usability would be great - instead of lots of explicit ignore lists, there would be a simple implementation that only looks at the fields that are trying to be set, and the references to unrelated products like Rancher would be removed. This is all a much cleaner design